A problem to train a model for Russian language

1. I took 5 million lines from Russian wikipedia dump (extracted text), create alphabet_ru.txt with the following text (all files in UTF-8):
абвгдеёжзийклмнопрстуфхцчьъшщыэюя

2. I trained a model:
~/JamSpell/build$ ./main/jamspell train alphabet_ru.txt ~/Downloads/xaa model_wiki_ru.bin
[info] loading text
[info] generating N-grams 0
[info] processed 0%
[info] generating keys
[info] ngrams1: 1592588
[info] ngrams2: 27563594
[info] ngrams3: 57626371
[info] total: 86782553
[info] generating perf hash
[info] finished, buckets: 108478199
[info] buckets filled

It looks like it was created without errors but when I tried to correct misspelled words it doesn't work:

import jamspell
corrector = jamspell.TSpellCorrector()
corrector.LoadLangModel('model_wiki_ru.bin')
corrector.FixFragment(u'Папа пощел погуоять в метро.')

corrector.GetCandidates([u'погуоять'], 0)
()

"пощел погуоять" weren't corrected! In the same time your small model correct these words!
I tried phrases with several completely corrupted words with zero effect, no correct, no suggestions

Where is my mistake?

I update the code, recompile jamspell from source. Here are results of tests:

`python -m pytest test_jamspell.py
================================================================== test session starts ==================================================================
platform linux2 -- Python 2.7.12, pytest-3.4.0, py-1.5.2, pluggy-0.6.0
rootdir: /home/antoly/JamSpell, inifile:
collected 2 items                                                                                                                                       

test_jamspell.py FF                                                                                                                               [100%]

======================================================================= FAILURES ========================================================================
_____________________________________________ test_evaluation[sherlockholmes.txt-alphabet_en.txt-expected0] _____________________________________________

sourceFile = 'sherlockholmes.txt', alphabetFile = 'test_data/alphabet_en.txt'
expected = (0.04519985057900635, 0.7005163511187608, 0.014246804944479363, 0.01363466567052671, 0.7676419965576592)

    @pytest.mark.parametrize('sourceFile,alphabetFile,expected', [
        ('sherlockholmes.txt', 'alphabet_en.txt', (0.04519985057900635, 0.7005163511187608, 0.014246804944479363,
                                                   0.01363466567052671, 0.7676419965576592)),
        ('kapitanskaya_dochka.txt', 'alphabet_ru.txt', (0.12330535829567463, 0.391304347826087, 0.03866565579984837,
                                                        0.05422853453841188, 0.4391304347826087)),
    ])
    def test_evaluation(sourceFile, alphabetFile, expected):
        alphabetFile = TEST_DATA + alphabetFile
        generate_dataset.generateDatasetTxt(TEST_DATA + sourceFile, TEMP)
        trainLangModel(TEMP_TRAIN, alphabetFile, TEMP_MODEL)
        results = evaluateJamspell(TEMP_MODEL, TEMP_TEST, alphabetFile)
>       assert results == expected
E       assert (0.1667911841...16137467, 0.0) == (0.04519985057...6419965576592)
E         At index 0 diff: 0.16679118416137467 != 0.04519985057900635
E         Use -v to get the full diff

test_jamspell.py:42: AssertionError
----------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------
[info] removing duplicates
[info] 9986 left
[info] shuffling
[info] saving train set
[info] saving test set
[info] done
----------------------------------------------------------------- Captured stderr call ------------------------------------------------------------------
[info] loading text
[info] generating N-grams 0
[info] generating keys
[info] ngrams1: 13077
[info] ngrams2: 55608
[info] ngrams3: 86329
[info] total: 155014
[info] generating perf hash
[info] finished, buckets: 193771
[info] buckets filled
__________________________________________ test_evaluation[kapitanskaya_dochka.txt-alphabet_ru.txt-expected1] ___________________________________________

sourceFile = 'kapitanskaya_dochka.txt', alphabetFile = 'test_data/alphabet_ru.txt'
expected = (0.12330535829567463, 0.391304347826087, 0.03866565579984837, 0.05422853453841188, 0.4391304347826087)

    @pytest.mark.parametrize('sourceFile,alphabetFile,expected', [
        ('sherlockholmes.txt', 'alphabet_en.txt', (0.04519985057900635, 0.7005163511187608, 0.014246804944479363,
                                                   0.01363466567052671, 0.7676419965576592)),
        ('kapitanskaya_dochka.txt', 'alphabet_ru.txt', (0.12330535829567463, 0.391304347826087, 0.03866565579984837,
                                                        0.05422853453841188, 0.4391304347826087)),
    ])
    def test_evaluation(sourceFile, alphabetFile, expected):
        alphabetFile = TEST_DATA + alphabetFile
        generate_dataset.generateDatasetTxt(TEST_DATA + sourceFile, TEMP)
        trainLangModel(TEMP_TRAIN, alphabetFile, TEMP_MODEL)
        results = evaluateJamspell(TEMP_MODEL, TEMP_TEST, alphabetFile)
>       assert results == expected
E       assert (0.2724338282...82763073, 0.0) == (0.12330535829...1304347826087)
E         At index 0 diff: 0.2724338282763073 != 0.12330535829567463
E         Use -v to get the full diff

test_jamspell.py:42: AssertionError
----------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------
[info] removing duplicates
[info] 802 left
[info] shuffling
[info] saving train set
[info] saving test set
[info] done
----------------------------------------------------------------- Captured stderr call ------------------------------------------------------------------
[info] loading text
[info] generating N-grams 0
[info] generating keys
[info] ngrams1: 8499
[info] ngrams2: 23588
[info] ngrams3: 28200
[info] total: 60287
[info] generating perf hash
[info] finished, buckets: 75367
[info] buckets filled`

--- a/jamspell/lang_model.cpp
+++ b/jamspell/lang_model.cpp
@@ -127,7 +127,7 @@ bool TLangModel::Train(const std::string& fileName, const std::string& alphabetF
     std::unordered_map<TGram2Key, TCount, TGram2KeyHash> grams2;
     std::unordered_map<TGram3Key, TCount, TGram3KeyHash> grams3;
 
-    std::cerr << "[info] generating N-grams " << sentences.size() << std::endl;
+    std::cerr << "[info] generating N-grams " << sentenceIds.size() << std::endl;
     uint64_t lastTime = GetCurrentTimeMs();
     size_t total = sentenceIds.size();
     for (size_t i = 0; i < total; ++i) {

--- a/jamspell/utils.cpp
+++ b/jamspell/utils.cpp
@@ -113,7 +113,7 @@ std::wstring UTF8ToWide(const std::string& text) {
     using boost::locale::conv::utf_to_utf;
     return utf_to_utf<wchar_t>(text.c_str(), text.c_str() + text.size());
 #else
-    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
+    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t, 0x10ffff, std::little_endian>> converter;
     return converter.from_bytes(text);
 #endif
 }
@@ -123,7 +123,7 @@ std::string WideToUTF8(const std::wstring& text) {
     using boost::locale::conv::utf_to_utf;
     return utf_to_utf<char>(text.c_str(), text.c_str() + text.size());
 #else
-    using convert_type = std::codecvt_utf8<wchar_t>;
+    using convert_type = std::codecvt_utf8<wchar_t, 0x10ffff, std::little_endian>;
     std::wstring_convert<convert_type, wchar_t> converter;
     return converter.to_bytes(text);
 #endif

